File: src/generators/domain.ts
import type { GeneratedFile, Generator, GeneratorConfig, PromptQuestion } from './types.js'

function toSlug(value: string): string {
  return value
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
}

function toTitleCase(value: string): string {
  return value
    .split(/[-_\s]+/)
    .filter(Boolean)
    .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
    .join(' ')
}

export class DomainGenerator implements Generator {
  readonly type = 'domain' as const

  getPromptQuestions(): PromptQuestion[] {
    return [
      {
        key: 'description',
        label: 'Domain skill description',
        type: 'text',
        required: false,
      },
    ]
  }

  async generate(config: GeneratorConfig): Promise<GeneratedFile[]> {
    const slug = toSlug(config.name)
    const title = toTitleCase(slug || config.name)
    const description = String(config.answers?.description ?? config.description ?? `Domain knowledge for ${title.toLowerCase()} work.`)

    const content = `---
kind: domain-skill
name: ${slug || 'new-domain'}
description: ${description}
applies_to:
  - scout
  - planner
  - builder
  - reviewer
knowledge_areas:
  - ${slug || 'domain-area'}
allowed_tools:
  - Read
  - Grep
  - Glob
  - Edit
  - Write
model_hint: sonnet
---

# ${title} Domain Skill

Use this skill to inject project-specific domain knowledge for ${title.toLowerCase()} tasks.

- surface domain constraints early
- prefer explicit contracts and invariants
- document risks, assumptions, and edge cases
`

    return [
      {
        path: `.ai/orchestration/skills/domains/${slug || 'new-domain'}.md`,
        content,
      },
    ]
  }
}

File: packages/ai-setup-go/internal/generator/domain.go
package generator

import (
	"fmt"

	"github.com/ricardoborges-teachable/ai-setup/internal/types"
)

// DomainGenerator generates domain skill files.
// Ported from src/generators/domain.ts.
type DomainGenerator struct{}

func (g *DomainGenerator) Type() types.ArtifactType { return types.ArtifactTypeDomain }

func (g *DomainGenerator) GetPromptQuestions() []PromptQuestion {
	return []PromptQuestion{
		{
			Key:      "description",
			Label:    "Domain skill description",
			Type:     "text",
			Required: false,
		},
	}
}

func (g *DomainGenerator) Generate(config GeneratorConfig) ([]GeneratedFile, error) {
	slug := ToSlug(config.Name)
	title := ToTitleCase(slug)
	if title == "" {
		title = config.Name
	}

	description := getAnswer(config.Answers, "description", "")
	if description == "" {
		description = config.Description
	}
	if description == "" {
		description = fmt.Sprintf("Domain knowledge for %s work.", stringsToLower(title))
	}

	domainSlug := slug
	if domainSlug == "" {
		domainSlug = "new-domain"
	}

	content := fmt.Sprintf(`---
kind: domain-skill
name: %s
description: %s
applies_to:
  - scout
  - planner
  - builder
  - reviewer
knowledge_areas:
  - %s
allowed_tools:
  - Read
  - Grep
  - Glob
  - Edit
  - Write
model_hint: sonnet
---

# %s Domain Skill

Use this skill to inject project-specific domain knowledge for %s tasks.

- surface domain constraints early
- prefer explicit contracts and invariants
- document risks, assumptions, and edge cases
`,
		domainSlug,
		description,
		domainSlug,
		title,
		stringsToLower(title),
	)

	return []GeneratedFile{
		{
			Path:    fmt.Sprintf(".ai/orchestration/skills/domains/%s.md", domainSlug),
			Content: content,
		},
	}, nil
}

// stringsToLower is a simple ASCII lowercase helper.
func stringsToLower(s string) string {
	result := make([]byte, len(s))
	for i, r := range s {
		if r >= 'A' && r <= 'Z' {
			result[i] = byte(r + 32)
		} else {
			result[i] = byte(r)
		}
	}
	return string(result)
}
